resolvers/onchain.go

```diff
@@ -0,0 +1,378 @@
+				msg)
+		}
+	} else {
+		if onchainRevStatus.genesisState == nil {
+			return out, errors.New(
+				"genesis state is not specified in OnChainCredentialStatus ID")
+		}
+		resp, err = contractCaller.GetRevocationStatusByIdAndState(
+			contractOpts,
+			issuerID.BigInt(), onchainRevStatus.genesisState,
+			onchainRevStatus.revNonce)
+		if err != nil {
+			return out, fmt.Errorf(
+				"GetRevocationProof smart contract call [GetRevocationStatusByIdAndState]: %s",
+				err.Error())
+		}
+	}
+
+	return toRevocationStatus(resp)
+}
+
+func newOnchainRevStatusFromURI(stateID string) (onChainRevStatus, error) {
+	var s onChainRevStatus
+
+	uri, err := url.Parse(stateID)
+	if err != nil {
+		return s, errors.New("OnChainCredentialStatus ID is not a valid URI")
+	}
+
+	contract := uri.Query().Get("contractAddress")
+	if contract == "" {
+		return s, errors.New("OnChainCredentialStatus contract address is empty")
+	}
+
+	contractParts := strings.Split(contract, ":")
+	if len(contractParts) != 2 {
+		return s, errors.New(
+			"OnChainCredentialStatus contract address is not valid")
+	}
+
+	s.chainID, err = newChainIDFromString(contractParts[0])
+	if err != nil {
+		return s, err
+	}
+	s.contractAddress = common.HexToAddress(contractParts[1])
+
+	revocationNonce := uri.Query().Get("revocationNonce")
+	if revocationNonce == "" {
+		return s, errors.New("revocationNonce is empty in OnChainCredentialStatus ID")
+	}
+
+	s.revNonce, err = strconv.ParseUint(revocationNonce, 10, 64)
+	if err != nil {
+		return s, errors.New("revocationNonce is not a number in OnChainCredentialStatus ID")
+	}
+
+	// state may be nil if params is absent in query
+	s.genesisState, err = newIntFromHexQueryParam(uri, "state")
+	if err != nil {
+		return s, err
+	}
+
+	return s, nil
+}
+
+func newChainIDFromString(in string) (core.ChainID, error) {
+	var chainID uint64
+	var err error
+	if strings.HasPrefix(in, "0x") ||
+		strings.HasPrefix(in, "0X") {
+		chainID, err = strconv.ParseUint(in[2:], 16, 64)
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		chainID, err = strconv.ParseUint(in, 10, 64)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return core.ChainID(chainID), nil
+}
+
+// newIntFromHexQueryParam search for query param `paramName`, parse it
+// as hex string of LE bytes of *big.Int. Return nil if param is not found.
+func newIntFromHexQueryParam(uri *url.URL, paramName string) (*big.Int, error) {
+	stateParam := uri.Query().Get(paramName)
+	if stateParam == "" {
+		return nil, nil
+	}
+
+	stateParam = strings.TrimSuffix(stateParam, "0x")
+	stateBytes, err := hex.DecodeString(stateParam)
+	if err != nil {
+		return nil, err
+	}
+
+	return newIntFromBytesLE(stateBytes), nil
+}
+
+func newIntFromBytesLE(bs []byte) *big.Int {
+	return new(big.Int).SetBytes(utils.SwapEndianness(bs))
+}
+
+func stateContractHasID(ctx context.Context, stateAddr common.Address, ethClient *ethclient.Client, id *core.ID) (bool, error) {
```

Comment:
It is very psycho function :guardsman:
Why do you use mutex here and have duplicate code to get id, but one getter uses Readmutex but another uses FullLockMutext
```go
idsInStateContractLock.RLock()
	ok := idsInStateContract[*id]
	idsInStateContractLock.RUnlock()
	if ok {
		return ok, nil
	}

	idsInStateContractLock.Lock()
	defer idsInStateContractLock.Unlock()

	ok = idsInStateContract[*id]
	if ok {
		return ok, nil
	}
```

Reply:
@olomix